datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaConnector.java

```diff
@@ -147,32 +156,48 @@ public synchronized void onAssignmentChange(List<DatastreamTask> tasks) {
       Set<DatastreamTask> toCancel = new HashSet<>(_runningTasks.keySet());
       toCancel.removeAll(tasks);
 
-      for (DatastreamTask task : toCancel) {
-        ConnectorTaskEntry connectorTaskEntry = _runningTasks.remove(task);
-        // Spawn a separate thread to attempt stopping the connectorTask. The connectorTask will be canceled if it
-        // does not stop within a certain amount of time. This will force cleanup of connectorTasks which take too long
-        // to stop, or are stuck indefinitely. A separate thread is spawned to handle this because the Coordinator
-        // requires that onAssignmentChange() must complete quickly, and will kill the assignment threads if they take
-        // too long.
-        _shutdownExecutorService.submit(() -> {
-          stopTask(task, connectorTaskEntry);
-        });
+      if (toCancel.size() > 0) {
+        // Mark the connector task as stopped so that, in case stopping the task here fails for any reason in
+        // restartDeadTasks the task is not restarted
+        synchronized (_toStopTasks) {
+          toCancel.stream().forEach(task -> {
+            _toStopTasks.put(task, _runningTasks.get(task));
+            _runningTasks.remove(task);
+          });
+        }
+        stopUnassignedTasks();
       }
 
-      for (DatastreamTask task : tasks) {
-        ConnectorTaskEntry connectorTaskEntry = _runningTasks.get(task);
-        if (connectorTaskEntry != null) {
-          AbstractKafkaBasedConnectorTask kafkaBasedConnectorTask = connectorTaskEntry.getConnectorTask();
-          kafkaBasedConnectorTask.checkForUpdateTask(task);
-          // Make sure to replace the DatastreamTask with most up to date info
-          // This is necessary because DatastreamTaskImpl.hashCode() does not take into account all the
-          // fields/properties of the DatastreamTask (e.g. dependencies).
-          _runningTasks.remove(task);
-          _runningTasks.put(task, connectorTaskEntry);
-          continue; // already running
+      boolean toCallRestartDeadTasks = false;
+      // This is the only place where we synchronize on both _toStopTasks and _runningTasks. But if such a case
+      // arises elsewhere the order of synchronization should be _runningTasks and then _toStopTasks.
```

Comment:
In the previous code block you are also accessing the lock for `_toStopTasks`, so this comment should be moved  up and modified to talk about accessing both in this function.
Is it really necessary to leave a  comment here since you've already left a comment where these are defined? In the future if someone adds another block where both are accessed, they may miss updating/removing this comment.

also nit to reword in case you decide to keep it:
"We only synchronize both _runningTasks and _toStopTasks in this function. If these need to be synchronized together in other parts of the code, please follow the correct lock ordering semantics."